File: app/workbench/imports.ts
import { csvObjects } from "../../packages/connectors/src/csv";
import {
  SCHEMA_VERSION,
  parseShockScript,
  sealSnapshot,
  sha256Text,
  verifySnapshot,
  type GraphSnapshot,
  type GraphSnapshotDraft,
  type ShockScenario,
  type WorldEdge,
  type WorldNode,
} from "../../packages/core/src/index";

const maximumGraphBytes = 5_000_000;
const maximumNodes = 10_000;
const maximumEdges = 50_000;

function slug(value: string): string {
  const output = value
    .normalize("NFKD")
    .toLowerCase()
    .replace(/[^a-z0-9._-]+/g, "-")
    .replace(/^[-._]+|[-._]+$/g, "")
    .slice(0, 80);
  return output || "item";
}

function unitInterval(value: unknown, fallback: number, path: string): number {
  if (value === undefined || value === null || value === "") return fallback;
  const parsed = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(parsed) || parsed < 0 || parsed > 1) {
    throw new RangeError(`${path} must be a finite number between 0 and 1.`);
  }
  return parsed;
}

interface RawNode {
  id: string;
  label?: string;
  kind?: string;
  [key: string]: unknown;
}

interface RawEdge {
  source: string;
  target: string;
  relation?: string;
  weight?: unknown;
  lower?: unknown;
  upper?: unknown;
  [key: string]: unknown;
}

async function normalizedSnapshot(
  text: string,
  title: string,
  rawNodes: RawNode[],
  rawEdges: RawEdge[],
  contentType: string,
): Promise<GraphSnapshot> {
  if (rawNodes.length === 0 || rawNodes.length > maximumNodes) {
    throw new RangeError(`Use 1-${maximumNodes.toLocaleString()} nodes.`);
  }
  if (rawEdges.length > maximumEdges) {
    throw new RangeError(`Use at most ${maximumEdges.toLocaleString()} edges.`);
  }
  const cutoff = new Date().toISOString();
  const sourceId = "source:user-provided:graph";
  const idByRaw = new Map<string, string>();
  const used = new Set<string>();
  const nodes: WorldNode[] = rawNodes.map((item) => {
    if (idByRaw.has(item.id)) {
      throw new TypeError(`Duplicate node id: ${item.id}`);
    }
    let id = `node:user:${slug(item.id)}`;
    let suffix = 2;
    while (used.has(id)) {
      id = `node:user:${slug(item.id)}-${suffix}`;
      suffix += 1;
    }
    used.add(id);
    idByRaw.set(item.id, id);
    const kind = [
      "country", "region", "product", "industry", "legal_entity", "facility",
      "port", "route", "security", "fund", "medicine", "policy", "event", "metric",
    ].includes(String(item.kind)) ? item.kind as WorldNode["kind"] : "metric";
    return {
      id,
      kind,
      label: String(item.label ?? item.id),
      description: "Imported user-provided node; not independently verified by CascadeLens.",
      validFrom: cutoff,
      observedAt: cutoff,
      properties: {
        bufferShare: unitInterval(
          item.bufferShare ?? item.buffer_share,
          0,
          `Node ${item.id} bufferShare`,
        ),
        criticality: unitInterval(item.criticality, 1, `Node ${item.id} criticality`),
        factualStatus: "user_provided_unverified",
      },
      evidence: {
        grade: "MODEL_INFERRED",
        confidence: 0.25,
        sourceIds: [sourceId],
        reviewStatus: "not_required",
      },
    };
  });
  const edges: WorldEdge[] = rawEdges.map((item, index) => {
    const from = idByRaw.get(item.source);
    const to = idByRaw.get(item.target);
    if (!from || !to) throw new TypeError(`Edge ${index + 1} references an unknown node.`);
    const value = unitInterval(item.weight, 0.5, `Edge ${index + 1} weight`);
    const lower = unitInterval(item.lower, value, `Edge ${index + 1} lower`);
    const upper = unitInterval(item.upper, value, `Edge ${index + 1} upper`);
    if (lower > value || value > upper) {
      throw new RangeError(`Edge ${index + 1} requires lower <= weight <= upper.`);
    }
    const relations = [
      "trades_to", "supplies", "inputs_to", "depends_on", "owns", "holds",
      "located_in", "connects_to", "substitute_for", "regulated_by", "exposed_to",
      "disclosed_relation",
    ];
    return {
      id: `edge:user:${String(index + 1).padStart(6, "0")}`,
      from,
      to,
      relation: relations.includes(String(item.relation))
        ? item.relation as WorldEdge["relation"]
        : "depends_on",
      weight: { value, lower, upper, unit: "share" },
      validFrom: cutoff,
      observedAt: cutoff,
      properties: {
        eligibleForPrimaryEstimate: false,
        factualStatus: "user_provided_unverified",
      },
      evidence: {
        grade: "MODEL_INFERRED",
        confidence: 0.25,
        sourceIds: [sourceId],
        reviewStatus: "not_required",
      },
    };
  });
  const draft: GraphSnapshotDraft = {
    schemaVersion: SCHEMA_VERSION,
    snapshotId: `snapshot:user:${(await sha256Text(text)).slice(0, 20)}`,
    title,
    decisionCutoff: cutoff,
    generatedAt: cutoff,
    nodes,
    edges,
    sources: [{
      id: sourceId,
      title,
      publisher: "User provided",
      uri: "https://github.com/limingrui679-design/CascadeLens/blob/main/docs/tutorials/02_bring_your_own_graph.md",
      retrievedAt: cutoff,
      availableAt: cutoff,
      publishedAt: cutoff,
      sha256: await sha256Text(text),
      contentType,
      artifactKind: "raw_snapshot",
      digestScope: "exact_bytes",
      bytes: new TextEncoder().encode(text).byteLength,
      role: "input",
      license: {
        mode: "user_provided",
        name: "User-provided data; user controls terms",
        termsUri: "https://github.com/limingrui679-design/CascadeLens/blob/main/docs/DATA_LICENSES.md",
        notes: "CascadeLens does not infer redistribution rights for imported data.",
      },
    }],
  };
  return sealSnapshot(draft);
}

function simpleJson(value: unknown): { nodes: RawNode[]; edges: RawEdge[]; title: string } {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError("JSON graph must be an object.");
  }
  const document = value as Record<string, unknown>;
  if (!Array.isArray(document.nodes) || !Array.isArray(document.edges)) {
    throw new TypeError("Simple JSON needs nodes and edges arrays.");
  }
  const nodes = document.nodes.map((item, index): RawNode => {
    if (typeof item === "string") return { id: item, label: item };
    if (!item || typeof item !== "object" || Array.isArray(item)) {
      throw new TypeError(`nodes[${index}] must be a string or object.`);
    }
    const record = item as Record<string, unknown>;
    const id = record.id ?? record.name;
    if (typeof id !== "string" || id.trim() === "") {
      throw new TypeError(`nodes[${index}] needs id or name.`);
    }
    return { ...record, id } as RawNode;
  });
  const edges = document.edges.map((item, index): RawEdge => {
    if (!item || typeof item !== "object" || Array.isArray(item)) {
      throw new TypeError(`edges[${index}] must be an object.`);
    }
    const record = item as Record<string, unknown>;
    const source = record.source ?? record.from;
    const target = record.target ?? record.to;
    if (typeof source !== "string" || typeof target !== "string") {
      throw new TypeError(`edges[${index}] needs source/from and target/to.`);
    }
    return { ...record, source, target } as RawEdge;
  });
  return { nodes, edges, title: String(document.title ?? "Imported JSON graph") };
}

function graphMl(text: string): { nodes: RawNode[]; edges: RawEdge[] } {
  if (/<!\s*(?:DOCTYPE|ENTITY)\b/i.test(text)) {
    throw new TypeError("GraphML document type and entity declarations are not supported.");
  }
  const document = new DOMParser().parseFromString(text, "application/xml");
  const error = document.querySelector("parsererror");
  if (error) throw new SyntaxError("GraphML could not be parsed.");
  const keyNames = new Map<string, string>();
  document.querySelectorAll("key").forEach((key) => {
    const id = key.getAttribute("id");
    if (id) keyNames.set(id, key.getAttribute("attr.name") ?? id);
  });
  const data = (element: Element) => {
    const output: Record<string, string> = {};
    Array.from(element.children).filter((child) => child.localName === "data").forEach((child) => {
      const key = child.getAttribute("key") ?? "value";
      output[keyNames.get(key) ?? key] = child.textContent?.trim() ?? "";
    });
    return output;
  };
  const nodes = Array.from(document.getElementsByTagNameNS("*", "node")).map((node) => {
    const id = node.getAttribute("id");
    if (!id) throw new TypeError("GraphML node is missing id.");
    const attrs = data(node);
    return { id, label: attrs.label ?? attrs.name ?? id, ...attrs };
  });
  const graph = document.getElementsByTagNameNS("*", "graph").item(0);
  if (!graph) throw new TypeError("GraphML document contains no graph element.");
  const undirected = graph.getAttribute("edgedefault") === "undirected";
  const edges = Array.from(document.getElementsByTagNameNS("*", "edge")).flatMap((edge) => {
    const source = edge.getAttribute("source");
    const target = edge.getAttribute("target");
    if (!source || !target) throw new TypeError("GraphML edge is missing source or target.");
    const attributes = data(edge);
    const directed = edge.getAttribute("directed");
    const expand = directed === "false" || (directed !== "true" && undirected);
    return expand
      ? [{ source, target, ...attributes }, { source: target, target: source, ...attributes }]
      : [{ source, target, ...attributes }];
  });
  return { nodes, edges };
}

export async function importGraphFile(file: File): Promise<GraphSnapshot> {
  if (file.size > maximumGraphBytes) throw new RangeError("Graph file exceeds the 5 MB browser limit.");
  const text = await file.text();
  const suffix = file.name.toLowerCase().split(".").at(-1);
  if (suffix === "json") {
    const value = JSON.parse(text) as unknown;
    if (value && typeof value === "object" && !Array.isArray(value) && "contentDigest" in value) {
      const snapshot = value as GraphSnapshot;
      const issues = await verifySnapshot(snapshot);
      if (issues.some((issue) => issue.severity === "error")) {
        throw new TypeError(`Invalid WorldGraph: ${issues.map((issue) => `${issue.path}: ${issue.message}`).join("; ")}`);
      }
      return snapshot;
    }
    const graph = simpleJson(value);
    return normalizedSnapshot(text, graph.title, graph.nodes, graph.edges, file.type || "application/json");
  }
  if (suffix === "csv") {
    const rows = csvObjects(text);
    const nodeById = new Map<string, RawNode>();
    const edges = rows.map((row, index): RawEdge => {
      const source = row.source ?? row.from;
      const target = row.target ?? row.to;
      if (!source || !target) throw new TypeError(`CSV row ${index + 2} needs source/target or from/to.`);
      nodeById.set(source, { id: source, label: row.source_label || source });
      nodeById.set(target, { id: target, label: row.target_label || target });
      return { ...row, source, target };
    });
    return normalizedSnapshot(text, "Imported CSV dependency graph", [...nodeById.values()], edges, file.type || "text/csv");
  }
  if (suffix === "graphml" || suffix === "xml") {
    const graph = graphMl(text);
    return normalizedSnapshot(text, "Imported GraphML dependency graph", graph.nodes, graph.edges, file.type || "application/graphml+xml");
  }
  throw new TypeError("Choose a .json, .csv, or .graphml graph file.");
}

export async function importScenarioFile(file: File): Promise<ShockScenario> {
  if (file.size > 1_000_000) throw new RangeError("ShockScript exceeds the 1 MB limit.");
  return parseShockScript(await file.text());
}

export function starterScenario(snapshot: GraphSnapshot): ShockScenario {
  const first = snapshot.nodes[0];
  const start = new Date(new Date(snapshot.decisionCutoff).getTime() + 86_400_000).toISOString();
  return {
    schemaVersion: SCHEMA_VERSION,
    scenarioId: "user-graph-stress",
    title: "User-provided graph stress",
    summary: "A scenario-only starter generated from a user-provided graph.",
    classification: "synthetic_stress",
    decisionCutoff: snapshot.decisionCutoff,
    graphSnapshotId: snapshot.snapshotId,
    shocks: [{
      id: "shock:user-graph:primary",
      label: `Stress ${first.label}`,
      target: { ids: [first.id] },
      operation: "reduce_supply",
      magnitude: 0.5,
      unit: "share",
      startsAt: start,
      rationale: "User-selected scenario parameter; not an observation.",
      sourceIds: [],
    }],
    propagation: {
      engine: "dependency_cascade",
      transmission: 0.75,
      maxIterations: 100,
      tolerance: 1e-9,
      horizonsDays: [7, 30, 90],
      bounds: ["lower", "central", "upper"],
    },
    interventions: [],
    objectives: [{ id: "objective:user:risk", metric: "residual_impact", sense: "minimize" }],
    constraints: {},
    limitations: [
      "User-provided topology and weights are not independently validated.",
      "Outputs are scenarios, not forecasts, causal estimates, or realized losses.",
    ],
  };
}

File: src/cascadelens/imports.py
"""Bring-your-own-graph adapters for JSON, CSV, GraphML, and NetworkX."""

from __future__ import annotations

import csv
import io
import json
import math
import re
import xml.etree.ElementTree as ET
from datetime import UTC, datetime
from pathlib import Path
from typing import Any, Protocol

from .canonical import sha256_bytes
from .temporal import isoformat
from .validation import NODE_KINDS, RELATION_KINDS, seal_snapshot, verify_snapshot
from .version import SCHEMA_VERSION

MAX_GRAPH_BYTES = 25_000_000
MAX_NODES = 100_000
MAX_EDGES = 500_000
_SAFE = re.compile(r"[^a-z0-9._-]+")


class NetworkXLike(Protocol):
    def nodes(self, data: bool = False): ...

    def edges(self, data: bool = False, keys: bool = False): ...

    def is_multigraph(self) -> bool: ...


def _slug(value: Any) -> str:
    text = _SAFE.sub("-", str(value).strip().lower()).strip("-._")
    if not text:
        text = sha256_bytes(str(value).encode("utf-8"))[:16]
    return text[:80]


def _now() -> str:
    return isoformat(datetime.now(UTC).replace(microsecond=0))


def _float(value: Any, fallback: float) -> float:
    if value is None or value == "":
        return fallback
    try:
        parsed = float(value)
    except (TypeError, ValueError) as error:
        raise ValueError(f"Expected a numeric value between 0 and 1, received {value!r}") from error
    if not math.isfinite(parsed) or not 0 <= parsed <= 1:
        raise ValueError(f"Expected a finite numeric value between 0 and 1, received {value!r}")
    return parsed


def _source_record(raw: bytes, cutoff: str, title: str) -> dict[str, Any]:
    return {
        "id": "source:user-provided:graph",
        "title": title,
        "publisher": "User provided",
        "uri": "https://github.com/limingrui679-design/CascadeLens/blob/main/docs/tutorials/02_bring_your_own_graph.md",
        "retrievedAt": cutoff,
        "availableAt": cutoff,
        "publishedAt": cutoff,
        "sha256": sha256_bytes(raw),
        "contentType": "application/octet-stream",
        "artifactKind": "raw_snapshot",
        "digestScope": "exact_bytes",
        "bytes": len(raw),
        "role": "input",
        "license": {
            "mode": "user_provided",
            "name": "User-provided data; user controls terms",
            "termsUri": "https://github.com/limingrui679-design/CascadeLens/blob/main/docs/DATA_LICENSES.md",
            "notes": "CascadeLens does not infer redistribution rights for imported data.",
        },
    }


def _node(node_id: str, label: str, attrs: dict[str, Any], cutoff: str) -> dict[str, Any]:
    kind = str(attrs.get("kind", "metric"))
    if kind not in NODE_KINDS:
        kind = "metric"
    criticality = _float(attrs.get("criticality", 1), 1)
    buffer_share = _float(attrs.get("bufferShare", attrs.get("buffer_share", 0)), 0)
    properties = {
        key: value
        for key, value in attrs.items()
        if key not in {"id", "label", "kind", "criticality", "bufferShare", "buffer_share"}
        and isinstance(value, (str, int, float, bool))
    }
    properties.update(
        {
            "criticality": criticality,
            "bufferShare": buffer_share,
            "factualStatus": "user_provided_unverified",
        }
    )
    return {
        "id": node_id,
        "kind": kind,
        "label": label,
        "description": "Imported user-provided node; not independently verified by CascadeLens.",
        "validFrom": cutoff,
        "observedAt": cutoff,
        "properties": properties,
        "evidence": {
            "grade": "MODEL_INFERRED",
            "confidence": 0.25,
            "sourceIds": ["source:user-provided:graph"],
            "reviewStatus": "not_required",
        },
    }


def _edge(
    edge_id: str,
    source_id: str,
    target_id: str,
    attrs: dict[str, Any],
    cutoff: str,
) -> dict[str, Any]:
    relation = str(attrs.get("relation", "depends_on"))
    if relation not in RELATION_KINDS:
        relation = "depends_on"
    value = _float(attrs.get("weight", attrs.get("value", 0.5)), 0.5)
    lower = _float(attrs.get("lower", value), value)
    upper = _float(attrs.get("upper", value), value)
    if not lower <= value <= upper:
        raise ValueError("Edge uncertainty requires lower <= weight <= upper")
    properties = {
        key: item
        for key, item in attrs.items()
        if key
        not in {
            "source",
            "target",
            "from",
            "to",
            "id",
            "weight",
            "value",
            "lower",
            "upper",
            "relation",
        }
        and isinstance(item, (str, int, float, bool))
    }
    properties.update(
        {
            "eligibleForPrimaryEstimate": False,
            "factualStatus": "user_provided_unverified",
        }
    )
    return {
        "id": edge_id,
        "from": source_id,
        "to": target_id,
        "relation": relation,
        "weight": {"value": value, "lower": lower, "upper": upper, "unit": "share"},
        "validFrom": cutoff,
        "observedAt": cutoff,
        "properties": properties,
        "evidence": {
            "grade": "MODEL_INFERRED",
            "confidence": 0.25,
            "sourceIds": ["source:user-provided:graph"],
            "reviewStatus": "not_required",
        },
    }


def _assemble(
    raw: bytes,
    title: str,
    nodes: list[tuple[Any, dict[str, Any]]],
    edges: list[tuple[Any, Any, dict[str, Any]]],
    *,
    decision_cutoff: str | None = None,
) -> dict[str, Any]:
    if not nodes:
        raise ValueError("Imported graph contains no nodes")
    if len(nodes) > MAX_NODES or len(edges) > MAX_EDGES:
        raise ValueError(f"Imported graph exceeds {MAX_NODES:,} nodes or {MAX_EDGES:,} edges")
    cutoff = decision_cutoff or _now()
    ids: dict[str, str] = {}
    normalized_nodes: list[dict[str, Any]] = []
    for raw_id, attrs in nodes:
        key = str(raw_id)
        if key in ids:
            raise ValueError(f"Imported graph contains duplicate node id {key!r}")
        base = f"node:user:{_slug(key)}"
        node_id = base
        counter = 2
        while node_id in ids.values():
            node_id = f"{base}-{counter}"
            counter += 1
        ids[key] = node_id
        normalized_nodes.append(_node(node_id, str(attrs.get("label", raw_id)), attrs, cutoff))
    normalized_edges: list[dict[str, Any]] = []
    for index, (raw_source, raw_target, attrs) in enumerate(edges, start=1):
        source_id = ids.get(str(raw_source))
        target_id = ids.get(str(raw_target))
        if source_id is None or target_id is None:
            raise ValueError(f"Edge {index} references a node missing from the node set")
        normalized_edges.append(
            _edge(f"edge:user:{index:06d}", source_id, target_id, attrs, cutoff)
        )
    draft = {
        "schemaVersion": SCHEMA_VERSION,
        "snapshotId": f"snapshot:user:{sha256_bytes(raw)[:20]}",
        "title": title,
        "decisionCutoff": cutoff,
        "generatedAt": cutoff,
        "nodes": normalized_nodes,
        "edges": normalized_edges,
        "sources": [_source_record(raw, cutoff, title)],
    }
    return seal_snapshot(draft)


def graph_from_json(
    value: str | bytes | dict[str, Any],
    *,
    decision_cutoff: str | None = None,
) -> dict[str, Any]:
    """Load a full WorldGraph or normalize a simple nodes/edges JSON object."""

    if isinstance(value, dict):
        document = value
        raw = json.dumps(value, ensure_ascii=False, sort_keys=True).encode("utf-8")
    else:
        raw = value.encode("utf-8") if isinstance(value, str) else value
        if len(raw) > MAX_GRAPH_BYTES:
            raise ValueError("Graph input exceeds the 25 MB safety limit")
        document = json.loads(raw)
    if len(raw) > MAX_GRAPH_BYTES:
        raise ValueError("Graph input exceeds the 25 MB safety limit")
    if not isinstance(document, dict):
        raise ValueError("JSON graph must be an object")
    if document.get("schemaVersion") == SCHEMA_VERSION and "contentDigest" in document:
        issues = verify_snapshot(document)
        if issues:
            rendered = "; ".join(issue.render() for issue in issues)
            raise ValueError(f"Invalid WorldGraph snapshot: {rendered}")
        return document
    raw_nodes = document.get("nodes")
    raw_edges = document.get("edges")
    if not isinstance(raw_nodes, list) or not isinstance(raw_edges, list):
        raise ValueError("Simple JSON graph requires nodes and edges arrays")
    nodes: list[tuple[Any, dict[str, Any]]] = []
    for index, item in enumerate(raw_nodes):
        if isinstance(item, str):
            nodes.append((item, {"label": item}))
        elif isinstance(item, dict):
            raw_id = item.get("id", item.get("name"))
            if raw_id is None:
                raise ValueError(f"nodes[{index}] requires id or name")
            nodes.append((raw_id, item))
        else:
            raise ValueError(f"nodes[{index}] must be a string or object")
    edges: list[tuple[Any, Any, dict[str, Any]]] = []
    for index, item in enumerate(raw_edges):
        if not isinstance(item, dict):
            raise ValueError(f"edges[{index}] must be an object")
        source = item.get("source", item.get("from"))
        target = item.get("target", item.get("to"))
        if source is None or target is None:
            raise ValueError(f"edges[{index}] requires source/from and target/to")
        edges.append((source, target, item))
    return _assemble(
        raw,
        str(document.get("title", "Imported JSON graph")),
        nodes,
        edges,
        decision_cutoff=decision_cutoff,
    )


def graph_from_csv(
    value: str | bytes,
    *,
    decision_cutoff: str | None = None,
) -> dict[str, Any]:
    """Normalize an edge-list CSV with source/target or from/to columns."""

    raw = value.encode("utf-8") if isinstance(value, str) else value
    if len(raw) > MAX_GRAPH_BYTES:
        raise ValueError("Graph input exceeds the 25 MB safety limit")
    text = raw.decode("utf-8-sig")
    reader = csv.DictReader(io.StringIO(text))
    if reader.fieldnames is None:
        raise ValueError("CSV graph requires a header row")
    fields = {field.strip().lower(): field for field in reader.fieldnames}
    source_field = fields.get("source", fields.get("from"))
    target_field = fields.get("target", fields.get("to"))
    if not source_field or not target_field:
        raise ValueError("CSV graph requires source/target or from/to columns")
    node_attrs: dict[str, dict[str, Any]] = {}
    edges: list[tuple[Any, Any, dict[str, Any]]] = []
    for index, row in enumerate(reader, start=2):
        source = (row.get(source_field) or "").strip()
        target = (row.get(target_field) or "").strip()
        if not source or not target:
            raise ValueError(f"CSV row {index} has an empty source or target")
        node_attrs.setdefault(source, {"label": row.get("source_label") or source})
        node_attrs.setdefault(target, {"label": row.get("target_label") or target})
        edges.append(
            (source, target, {key.strip().lower(): item for key, item in row.items() if key})
        )
    nodes = list(node_attrs.items())
    return _assemble(
        raw, "Imported CSV dependency graph", nodes, edges, decision_cutoff=decision_cutoff
    )


def graph_from_graphml(
    value: str | bytes,
    *,
    decision_cutoff: str | None = None,
) -> dict[str, Any]:
    """Normalize standard GraphML without executing extensions or fetching DTDs."""

    raw = value.encode("utf-8") if isinstance(value, str) else value
    if len(raw) > MAX_GRAPH_BYTES:
        raise ValueError("Graph input exceeds the 25 MB safety limit")
    if re.search(rb"<!\s*(?:DOCTYPE|ENTITY)\b", raw, re.IGNORECASE):
        raise ValueError("GraphML document type and entity declarations are not supported")
    root = ET.fromstring(raw)
    namespace = ""
    if root.tag.startswith("{"):
        namespace = root.tag.split("}", 1)[0] + "}"
    key_names: dict[str, str] = {}
    for key in root.findall(f"{namespace}key"):
        key_id = key.attrib.get("id")
        if key_id:
            key_names[key_id] = key.attrib.get("attr.name", key_id)
    graph = root.find(f"{namespace}graph")
    if graph is None:
        raise ValueError("GraphML document contains no graph element")

    def data(element: ET.Element) -> dict[str, Any]:
        attrs: dict[str, Any] = {}
        for child in element.findall(f"{namespace}data"):
            key = key_names.get(child.attrib.get("key", ""), child.attrib.get("key", "value"))
            attrs[key] = (child.text or "").strip()
        return attrs

    nodes: list[tuple[Any, dict[str, Any]]] = []
    for node in graph.findall(f"{namespace}node"):
        node_id = node.attrib.get("id")
        if node_id is None:
            raise ValueError("GraphML node is missing id")
        attrs = data(node)
        attrs.setdefault("label", attrs.get("name", node_id))
        nodes.append((node_id, attrs))
    edges: list[tuple[Any, Any, dict[str, Any]]] = []
    graph_is_undirected = graph.attrib.get("edgedefault") == "undirected"
    for edge in graph.findall(f"{namespace}edge"):
        source, target = edge.attrib.get("source"), edge.attrib.get("target")
        if source is None or target is None:
            raise ValueError("GraphML edge is missing source or target")
        attrs = data(edge)
        if "weight" in edge.attrib:
            attrs.setdefault("weight", edge.attrib["weight"])
        edges.append((source, target, attrs))
        directed = edge.attrib.get("directed")
        if directed == "false" or (directed != "true" and graph_is_undirected):
            edges.append((target, source, {**attrs, "graphml_undirected_expansion": True}))
    return _assemble(
        raw, "Imported GraphML dependency graph", nodes, edges, decision_cutoff=decision_cutoff
    )


def graph_from_networkx(
    graph: NetworkXLike,
    *,
    decision_cutoff: str | None = None,
) -> dict[str, Any]:
    """Normalize a NetworkX Graph, DiGraph, MultiGraph, or compatible object."""

    nodes = [(node_id, dict(attrs)) for node_id, attrs in graph.nodes(data=True)]
    edges: list[tuple[Any, Any, dict[str, Any]]] = []
    if graph.is_multigraph():
        for source, target, key, attrs in graph.edges(data=True, keys=True):
            payload = dict(attrs)
            payload.setdefault("networkx_key", str(key))
            edges.append((source, target, payload))
    else:
        edges = [(source, target, dict(attrs)) for source, target, attrs in graph.edges(data=True)]
    is_directed = getattr(graph, "is_directed", None)
    if callable(is_directed) and not is_directed():
        edges = edges + [
            (
                target,
                source,
                {**attrs, "networkx_undirected_expansion": True},
            )
            for source, target, attrs in edges
        ]
    raw = json.dumps(
        {"nodes": nodes, "edges": edges},
        ensure_ascii=False,
        sort_keys=True,
        default=str,
    ).encode("utf-8")
    return _assemble(
        raw, "Imported NetworkX dependency graph", nodes, edges, decision_cutoff=decision_cutoff
    )


def load_graph(path: str | Path, *, decision_cutoff: str | None = None) -> dict[str, Any]:
    source = Path(path)
    if not source.is_file() or source.is_symlink():
        raise ValueError(f"Expected a regular graph file: {source}")
    if source.stat().st_size > MAX_GRAPH_BYTES:
        raise ValueError("Graph input exceeds the 25 MB safety limit")
    raw = source.read_bytes()
    suffix = source.suffix.lower()
    if suffix == ".json":
        return graph_from_json(raw, decision_cutoff=decision_cutoff)
    if suffix == ".csv":
        return graph_from_csv(raw, decision_cutoff=decision_cutoff)
    if suffix in {".graphml", ".xml"}:
        return graph_from_graphml(raw, decision_cutoff=decision_cutoff)
    raise ValueError("Supported graph formats are .json, .csv, .graphml, and .xml")
